### PR TITLE
PD-274 Store IP in local storage

### PR DESCRIPTION
**Pull request by edwardupton**

### CODE REVIEW

File: src/segmentTracker/helpers.ts [hunk 1]
```diff
@@ -24,17 +25,22 @@ export const getContext = () => {
...
 			return resolve(standardContext);
 		}
+		const ipFromStorage = window.sessionStorage.getItem(sessionStorageKey);
+		if (ipFromStorage) {
+			return resolve({
+				...standardContext,
+				ip: ipFromStorage,
+			});
+		}
 		httpRequest
```
Comment (lyykfi) [inline src/segmentTracker/helpers.ts:35]: I would suggest to isolate the logic to a special helper like AnonymizedIpResolve for example.

**Review by lyykfi — CHANGES_REQUESTED** [review]

File: src/segmentTracker/helpers.ts [hunk 2]
```diff
@@ -176,3 +163,30 @@ export const callSegmentPage = (integrations: Record<string, any>) => {
...
+	return new Promise(resolve => {
+		const sessionStorageKey = 'littledata_ip_address';
+		const ipFromStorage = window.sessionStorage.getItem(sessionStorageKey);
+		if (ipFromStorage) {
+			return resolve({
+				...standardContext,
+				ip: ipFromStorage,
+			});
+		}
+		httpRequest
```
Comment (lyykfi) [inline src/segmentTracker/helpers.ts:177]: what do you think to use async/await here?

File: src/segmentTracker/helpers.ts [hunk 3]
```diff
@@ -176,3 +163,30 @@ export const callSegmentPage = (integrations: Record<string, any>) => {
 		);
 	});
 };
+
+const anonymizedIpResolve = (standardContext: object) => {
```
Comment (lyykfi) [inline src/segmentTracker/helpers.ts:167]: I would rather suggest don't have standardContext here.
I guess anonymizedIpResolve  should return only IP which was resolved.

What do you think?

**Review by lyykfi — CHANGES_REQUESTED** [review]

File: src/segmentTracker/helpers.ts (hunk 2 above)
Comment (edwardupton) [reply src/segmentTracker/helpers.ts:177]: We've had bad experiences of using async/await with webpack. Best not to

Comment (edwardupton) [reply src/segmentTracker/helpers.ts:177]: Making it webpack compatible with async/await doubles the size of the minified code and makes debugging much harder

Comment (lyykfi) [reply src/segmentTracker/helpers.ts:177]: That is fine. Though, what was a problem?

Comment (MihaiCernat) [reply src/segmentTracker/helpers.ts:177]: yes, it messes up badly sometimes

Comment (lyykfi) [reply src/segmentTracker/helpers.ts:177]: Agree, let skip for now.

File: src/segmentTracker/helpers.ts [hunk 4]
```diff
@@ -176,3 +163,24 @@ export const callSegmentPage = (integrations: Record<string, any>) => {
...
+		httpRequest
+			.getJSON('//api.ipify.org?format=json')
+			.then((response: any) => {
+				const { ip } = response;
+				const anonymizedIp = ip.replace(/([0-9a-z]+)$/, 0);
+				window.sessionStorage.setItem(sessionStorageKey, anonymizedIp);
+				resolve({ anonymizedIp });
+			})
+			.catch(() => {
+				resolve({});
```
Comment (lyykfi) [inline src/segmentTracker/helpers.ts:183]: should we use reject here?

**Comment by MihaiCernat** [comment]
> should we have reject here?

@lyykfi no, some tests are failing, checking now

File: test/segmentTracker/getContext.test.ts [hunk 5]
```diff
@@ -19,43 +19,41 @@ describe('getContext function', () => {
...
 		const context = await getContext();
-		//@ts-ignore
 		context.ip.should.equal('2001:0db8:85a3:0000:0000:8a2e:0370:0');
 	});
-	it('should call the right url to get the ip', async () => {
-		getJSON.resolves({ ip: '192.168.1.124' });
-		await getContext();
-		//@ts-ignore
-		getJSON.should.have.been.calledWith('//api.ipify.org?format=json');
+	it('does not call API if cached', async () => {
```
Comment (lyykfi) [inline test/segmentTracker/getContext.test.ts:53]: I think we should add a test for reject state.

**Review by lyykfi — CHANGES_REQUESTED** [review]

**Review by lyykfi — APPROVED** [review]

**Pull request merged**